### ISSUE TITLE
[HOLD for payment 2023-12-20] [$500] Task - When the receiver clicks on 'Copy to clipboard' for a task, the copied text shows incorrectly

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#29035]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Log in to User A's account.
2. Open User B's chat.
3. Go to the "Assign task" section.
4. Create a task by entering a title.
5. Open another browser and log in to User B's account.
6. Open User A's chat and click on “Copy to clipboard” of sent task
7. Paste it in the chat.
8. Notice that “Task for undefined” is copied instead of the actual task title
## Expected Result: 
When the receiver clicks on “Copy to clipboard” for a task, the copied text should display the task title, not “undefined”
## Actual Result: 
When the receiver clicks on “Copy to clipboard,” the copied text shows as "Task for undefined," which is incorrect

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.79-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/38435837/cdb38291-7cc9-487a-bd26-f1db110d9a90

https://github.com/Expensify/App/assets/38435837/56adb1e1-db44-4714-af40-41dd9f1c5d8f

https://github.com/Expensify/App/assets/38435837/1ca54a40-d0de-4c9b-9aa2-7011d574e164

https://github.com/Expensify/App/assets/38435837/fdf056cd-c05a-4c6d-b12a-463bca514a87

https://github.com/Expensify/App/assets/38435837/cf7e00a1-9a23-4aa0-afcc-218f40ce8f8f

https://github.com/Expensify/App/assets/38435837/0415ea07-68d6-4743-a3d7-1b33691cbc16

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1696612963386119

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01adb76d099d1a9033</li>
        <li>Upwork Job ID: 1729954370591846400</li>
        <li>Last Price Increase: 2023-11-29</li>
<li>Automatic offers: </li>
<ul>
<li>cubuspl42 | Reviewer | 27886951</li>
<li>paultsimura | Contributor | 27886953</li>
</ul></li>
    </ul>
</details>

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When a new task is created, the recipient gets "task for undefined" copied to the clipboard.

### What is the root cause of that problem?
When a new task is created, the task report behaves the same way as the thread report: it is not sent by Pusher to the recipient unless one of two happens:
- The recipient opens the task manually;
- The task was assigned to the recipient, not only shared with them;

Because of that, the task report is `undefined` for the recipient here:
https://github.com/Expensify/App/blob/9945d3c10ae2a6b51e1758ec56fdf6860bc90f30/src/libs/actions/Task.js#L922-L925

### What changes do you think we should make in order to solve the problem?
Since the task report is under the hood a child report of the "Task created" report action, even though the report is not sent by Pusher to the recipient, the original `reportAction` contains the field `childReportName`, which is updated by Pusher for both Sender and Recipient on every task name change:

<img width="1139" alt="image" src="https://github.com/Expensify/App/assets/12595293/f3376137-598b-4b66-8efd-b15fa529ea39">

After the recent changes, we use the `Task.getTaskReportActionMessage` in 2 places, I recommend making a broader refactoring:

#### First, refactor `Task.getTaskReportActionMessage`:

```js
/**
 * @param {string} actionName
 * @returns {string}
 */
function getTaskReportActionMessage(actionName) {
    switch (actionName) {
        case CONST.REPORT.ACTIONS.TYPE.TASKCOMPLETED:
            return Localize.translateLocal('task.messages.completed');
        case CONST.REPORT.ACTIONS.TYPE.TASKCANCELLED:
            return Localize.translateLocal('task.messages.canceled');
        case CONST.REPORT.ACTIONS.TYPE.TASKREOPENED:
            return Localize.translateLocal('task.messages.reopened');
        default:
            return Localize.translateLocal('task.task');
    }
}
```

We will use it as before – only for building the message for actions that are `isTaskAction == true`, and since it now builds only a static text, we don't need the 2nd and 3rd params.

Will update call of this function:

```diff
        onPress: (closePopover, {reportAction, selection}) => {
            const isTaskAction = ReportActionsUtils.isTaskAction(reportAction);
            const isCreateTaskAction = ReportActionsUtils.isCreatedTaskReportAction(reportAction);
            const isReportPreviewAction = ReportActionsUtils.isReportPreviewAction(reportAction);
            const message = _.last(lodashGet(reportAction, 'message', [{}]));
-           const reportID = lodashGet(reportAction, 'originalMessage.taskReportID', '').toString();
-           const messageHtml = isTaskAction || isCreateTaskAction ? Task.getTaskReportActionMessage(reportAction.actionName, reportID, isCreateTaskAction) : lodashGet(message, 'html', '');
+           const messageHtml = isTaskAction ? Task.getTaskReportActionMessage(reportAction.actionName) : lodashGet(message, 'html', '');
```

#### Second

For the actions that are `isCreateTaskAction == true`, we will add a separate branch of `if/else` (if needed, we can also localize the "task for" message):

```diff
            if (!isAttachment) {
                const content = selection || messageHtml;
                if (isReportPreviewAction) {
                    const iouReport = ReportUtils.getReport(ReportActionsUtils.getIOUReportIDFromReportActionPreview(reportAction));
                    const displayMessage = ReportUtils.getReportPreviewMessage(iouReport, reportAction);
                    Clipboard.setString(displayMessage);
                } else if (ReportActionsUtils.isModifiedExpenseAction(reportAction)) {
                    const modifyExpenseMessage = ReportUtils.getModifiedExpenseMessage(reportAction);
                    Clipboard.setString(modifyExpenseMessage);
                } else if (ReportActionsUtils.isMoneyRequestAction(reportAction)) {
                    const displayMessage = ReportUtils.getIOUReportActionDisplayMessage(reportAction);
                    Clipboard.setString(displayMessage);
+               } else if (isCreateTaskAction) {
+                   Clipboard.setString(`task for ${reportAction.childReportName}`);
                } else if (content) {
                    const parser = new ExpensiMark();
                    if (!Clipboard.canSetHtml()) {
                        Clipboard.setString(parser.htmlToMarkdown(content));
                    } else {
                        const plainText = parser.htmlToText(content);
                        Clipboard.setHtml(content, plainText);
                    }
                }
            }
```

Compared to simply removing the `isCreateTaskAction` check and using `message.html`, this approach is responsive to the task name changes.

### What alternative solutions did you explore? (Optional)

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When a new task is created, the recipient gets "task for undefined" copied to the clipboard.

### What is the root cause of that problem?
When a new task is created, the task report behaves the same way as the thread report: it is not sent by Pusher to the recipient unless one of two happens:
- The recipient opens the task manually;
- The task was assigned to the recipient, not only shared with them;

Because of that, the task report is `undefined` for the recipient here:
https://github.com/Expensify/App/blob/9945d3c10ae2a6b51e1758ec56fdf6860bc90f30/src/libs/actions/Task.js#L922-L925

### What changes do you think we should make in order to solve the problem?
Since the task report is under the hood a child report of the "Task created" report action, even though the report is not sent by Pusher to the recipient, the original `reportAction` contains the field `childReportName`, which **should be updated** by Pusher for both Sender and Recipient on every task name change:

<img width="1139" alt="image" src="https://github.com/Expensify/App/assets/12595293/f3376137-598b-4b66-8efd-b15fa529ea39">

After the recent changes, we use the `Task.getTaskReportActionMessage` in 2 places, I recommend making a broader refactoring:

#### Zero, make a Pusher change

#### First, refactor `Task.getTaskReportActionMessage`:

```js
/**
 * @param {string} actionName
 * @returns {string}
 */
function getTaskReportActionMessage(actionName) {
    switch (actionName) {
        case CONST.REPORT.ACTIONS.TYPE.TASKCOMPLETED:
            return Localize.translateLocal('task.messages.completed');
        case CONST.REPORT.ACTIONS.TYPE.TASKCANCELLED:
            return Localize.translateLocal('task.messages.canceled');
        case CONST.REPORT.ACTIONS.TYPE.TASKREOPENED:
            return Localize.translateLocal('task.messages.reopened');
        default:
            return Localize.translateLocal('task.task');
    }
}
```

We will use it as before – only for building the message for actions that are `isTaskAction == true`, and since it now builds only a static text, we don't need the 2nd and 3rd params.

Will update call of this function:

```diff
        onPress: (closePopover, {reportAction, selection}) => {
            const isTaskAction = ReportActionsUtils.isTaskAction(reportAction);
            const isCreateTaskAction = ReportActionsUtils.isCreatedTaskReportAction(reportAction);
            const isReportPreviewAction = ReportActionsUtils.isReportPreviewAction(reportAction);
            const message = _.last(lodashGet(reportAction, 'message', [{}]));
-           const reportID = lodashGet(reportAction, 'originalMessage.taskReportID', '').toString();
-           const messageHtml = isTaskAction || isCreateTaskAction ? Task.getTaskReportActionMessage(reportAction.actionName, reportID, isCreateTaskAction) : lodashGet(message, 'html', '');
+           const messageHtml = isTaskAction ? Task.getTaskReportActionMessage(reportAction.actionName) : lodashGet(message, 'html', '');
```

#### Second

For the actions that are `isCreateTaskAction == true`, we will add a separate branch of `if/else` (if needed, we can also localize the "task for" message):

```diff
            if (!isAttachment) {
                const content = selection || messageHtml;
                if (isReportPreviewAction) {
                    const iouReport = ReportUtils.getReport(ReportActionsUtils.getIOUReportIDFromReportActionPreview(reportAction));
                    const displayMessage = ReportUtils.getReportPreviewMessage(iouReport, reportAction);
                    Clipboard.setString(displayMessage);
                } else if (ReportActionsUtils.isModifiedExpenseAction(reportAction)) {
                    const modifyExpenseMessage = ReportUtils.getModifiedExpenseMessage(reportAction);
                    Clipboard.setString(modifyExpenseMessage);
                } else if (ReportActionsUtils.isMoneyRequestAction(reportAction)) {
                    const displayMessage = ReportUtils.getIOUReportActionDisplayMessage(reportAction);
                    Clipboard.setString(displayMessage);
+               } else if (isCreateTaskAction) {
+                   Clipboard.setString(`task for ${reportAction.childReportName}`);
                } else if (content) {
                    const parser = new ExpensiMark();
                    if (!Clipboard.canSetHtml()) {
                        Clipboard.setString(parser.htmlToMarkdown(content));
                    } else {
                        const plainText = parser.htmlToText(content);
                        Clipboard.setHtml(content, plainText);
                    }
                }
            }
```

**Compared to simply removing the `isCreateTaskAction` check and using `message.html`, this approach is responsive to the task name changes.**

To leverage this change, we also need to change the util function call here to just `Task.getTaskReportActionMessage(props.actionName)`:
https://github.com/Expensify/App/blob/d8d98cdc3401d5c7b71d21927f7cae633352e913/src/components/ReportActionItem/TaskAction.js#L20-L28

### What alternative solutions did you explore? (Optional)

**Comment by redpanda-bit** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Copied text pastes with undefined in the string.
### What is the root cause of that problem?
As mentioned in the other proposals, the function that copies the text relies on the reportAction to be in storage in order to copy the correct text. 
### What changes do you think we should make in order to solve the problem?

Given that we already have the content we need, we're showing it in the list item, we can move the logic we use in the list item ([TaskPreview.js](https://github.com/Expensify/App/blob/main/src/components/ReportActionItem/TaskPreview.js#L74C3-L80)) into a new function and use that function to do both:

a) render the content in the list item
<img width="551" alt="Screen Shot 2023-10-06 at 8 10 00 PM" src="https://github.com/Expensify/App/assets/25206487/cbba8f60-fa3b-4bde-88be-ab168ebfb91f">

b) render the content to be used in the clipboard function, with "Task for " prepended, lowercase or uppercase.
<img width="868" alt="call-same-function" src="https://github.com/Expensify/App/assets/25206487/c02ddd95-7fda-4cff-bae2-17c6b834be60">
#### Note: By retrieve I mean to just return, no api calls involved.

This approach copies and pastes "Task for " + <@mention if any> + taskTitle. Assuming the user wants to copy the entire text, including the mention.

### What alternative solutions did you explore? (Optional)

**Comment by esh-g** [comment]
Seems to be fixed in the latest staging and prod

https://github.com/Expensify/App/assets/77237602/9a34f4c4-8a86-4232-be2f-52ed9f45a222

**Comment by cubuspl42** [comment]
We have a consensus that we need backend changes here before we can move forward with front-end fixes.

C+ reviewed 🎀 👀 🎀

**Comment by cubuspl42** [comment]
@NikkiWines Would you review the "Zero, make a Pusher BE change" section in [this proposal](https://github.com/Expensify/App/issues/29035#issuecomment-1751461506)?

We need to figure out a plan for the next steps. Something that could work...

1. An internal engineer could add the necessary change to the back-end
2. We would then re-test the proposed solution(s) and resume the external process

Let me know what you think

**Comment by NikkiWines** [comment]
Sorry for the delay here, will give the BE proposal a review tomorrow!!

**Comment by NikkiWines** [comment]
:wave: Looks like the backend changes suggested here align with some proposed backend changes for [another issue](https://github.com/Expensify/App/issues/27018) related to task name changes / editing tasks, which is already underway (being handled by @marcaaron). 

Let's wait for those changes to be deployed and re-evaluate to determine what (if any) next steps are necessary to resolve this issue

**Comment by NikkiWines** [comment]
Ooh, although it looks like we've possibly decided to hold off on those changes  🤔 @thienlnam should we HOLD on this issue as well?

**Comment by thienlnam** [comment]
Ah this looks slightly different where we're just missing returning a piece of data from an Auth command that we should address. I can take the BE changes though, but won't be able to get to it until next week

**Comment by sonialiap** [comment]
@thienlnam booping the overdue label off of this issue and checking in if you think you'll have time this week to take a look at this?

**Comment by thienlnam** [comment]
Wrapping up some things from money2020 this week, though should have more time next week! Going to knock it on a weekly since it's not really a main flow atm

**Comment by thienlnam** [comment]
https://github.com/Expensify/App/issues/29035#issuecomment-1777708813 

Do have more time this week, but knocking back to a weekly still as it's not urgent

**Comment by thienlnam** [comment]
Spun up a couple PRs

**Comment by sonialiap** [comment]
@thienlnam how's it going?

**Comment by thienlnam** [comment]
Just put the last [PR](https://github.com/Expensify/Web-Expensify/pull/39532) in review

**Comment by sonialiap** [comment]
The [PR](https://github.com/Expensify/Web-Expensify/pull/39532) has merged 🎉  Are we done with this issue? @thienlnam

**Comment by paultsimura** [comment]
> Are we done with this issue?

The BE PRs were required to initiate the review of the proposals on this issue (it's still reproducible): https://github.com/Expensify/App/issues/29035#issuecomment-1752497582

@cubuspl42 could you please confirm?
Also, if you'd like to test [my proposal](https://github.com/Expensify/App/issues/29035#issuecomment-1751461506), feel free to use [this branch](https://github.com/paultsimura/Expensify/tree/fix/29035-task-clipboard-undefined).

**Comment by paultsimura** [comment]
@cubuspl42 from what came to my mind in terms of testing – yes, the BE changes fit the needs, and the FE changes fix the issue correctly.

After the fix:

https://github.com/Expensify/App/assets/12595293/2ef12f5f-1d82-4fdd-9608-0baf5467da1a

Before the fix:

https://github.com/Expensify/App/assets/12595293/2640d381-608e-44aa-be89-3e441d72367f

**Comment by cubuspl42** [comment]
C+ reviewed 🎀 👀 🎀

I approve the proposal by @paultsimura. It suggested the backend changes which were proved to help with solving this issue on the frontend side. The frontend changes in the presented Git branch also look good, with a note that we'll have to have localization in mind (i.e. Spanish translation).

**Comment by cubuspl42** [comment]
From the checklist:

> I verified any copy / text shown in the product is localized by adding it to src/languages/* files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)

If you modify some logic for displaying a message in a non-trivial way, this checkbox starts applying to you in the context of this message, in my opinion.

**Comment by cubuspl42** [comment]
>  I've fixed another issue

I just meant that this isn't really "another issue" in my interpretation. Solving unrelated issues in a PR for a specific issue might be not considered appropriate. Everything is fine, we can move on

**Comment by cubuspl42** [comment]
- The PR that introduced the bug has been identified. Link to the PR:
    -  N/A (This was a problem involving both a frontend and backend, it's difficult or impossible to point a single frontend PR)
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
    - N/A
- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
    - No need for additional discussion
- Determine if we should create a regression test for this bug.
    - Up to the QA team
- If we decide to create a regression test for the bug, please propose the regression test steps to ensure the same bug will not reach production again.
    - Log in as User A
    - Share a task with User B
    - Open another browser and log in as User B
    - Open the chat with User A chat and click on “Copy to clipboard” of the sent task, but don't open the task itself
    - Paste the copied text in the chat.
        - Verify that the text contains "task for ${title}" with the correct task title
        - Verify that the chat with User A has the "task for ${title}" with the correct task title in LHN
    - As User A, modify the Task title
    - As User B, click on “Copy to clipboard” of the task, and paste it in the chat.
        - Verify that the text contains "task for ${title}" with the updated task title
        - Verify that the chat with User A has the "task for ${title}" with the updated task title in LHN
    - Change language to Spanish
        - Verify that both LHN text and the one copied to the clipboard are translated into Spanish.

**Comment by sonialiap** [comment]
@cubuspl42  $500 - paid ✔️ 
@paultsimura $500 - paid ✔️ 
@ayazhussain79 report $50 - offer sent

**Comment by ayazhussain79** [comment]
@sonialiap Offer accepted, Thank you

**Comment by sonialiap** [comment]
Payment completed ✔️
